bindings/go/src/examples/perftest/perftest.go

```diff
@@ -312,45 +301,46 @@ func clientStart() error {
 		return err
 	}
 
-	perfTest.perThreadWorkers = make([]*UcpWorker, perfTestParams.numThreads)
-	perfTest.eps = make([]*UcpEp, perfTestParams.numThreads)
-	for i := 0; i < int(perfTestParams.numThreads); i += 1 {
-		initWorker(i)
-		if err := clientConnectWorker(i); err != nil {
-			return err
-		}
+	initWorker()
+	if err := clientConnectWorker(); err != nil {
+		return err
 	}
 
-	var totalDuration time.Duration = 0
+	var start time.Time
+	perfTest.quit = make(chan struct{})
+	perfTest.numCompletedRequests = int32(-perfTestParams.warmUpIter)
 	printHeader()
-	for i := -int(perfTestParams.warmUpIter); i < int(perfTestParams.numIterations); i += 1 {
-		perfTest.wg.Add(int(perfTestParams.numThreads))
-		for t := uint(0); t < perfTestParams.numThreads; t += 1 {
-			go clientThreadDoIter(i, t)
+	go printStatistics()
+	for {
+		threads := perfTestParams.numThreads
+		if perfTest.numCompletedRequests <= 0 {
+			start = time.Now()
+		} else if perfTest.numCompletedRequests == int32(perfTestParams.numIterations) {
+			break
+		} else if perfTest.numCompletedRequests > int32(perfTestParams.numIterations - perfTestParams.numThreads) {
+			threads = perfTestParams.numIterations - uint(perfTest.numCompletedRequests)
 		}
-		perfTest.wg.Wait()
-		var maxDuration time.Duration = 0
-		for _, threadDuration := range perfTest.completionTime {
-			if threadDuration > maxDuration {
-				maxDuration = threadDuration
-			}
+		perfTest.wg.Add(int(threads))
+		for t := uint(0); t < threads; t += 1 {
+			go clientThreadDoIter(t)
```

Comment:
so all threads use the same worker now?

Reply:
yes, from UCX perspective this supposed to correspond to ucx_perftest running single OS thread, number of go threads has meaning of window size

Reply:
do we also want a way to run multiple os threads?

Reply:
actually, the similarity to ucx_perftest isn’t exact. The Go runtime uses a many-to-many scheduler to map goroutines to OS threads, handling migrations and all other details in the background. This mapping is hidden from the Go program, so even though you can control the number of OS threads with GOMAXPROCS, maintaining a one worker per thread setup doesn’t seem practical